### ISSUE TITLE
Data::Dumper fails to escape bless class name

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8480]
Migrated from [rt.perl.org#39420](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=39420) (status was 'resolved')

Searchable as RT39420$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



$ cat t0\.pl
use Data&#8203;::Dumper;
print Dumper\(bless\(\{\}\, "a'b"\)\);
$ perl t0\.pl
$VAR1 = bless\( \{\}\, 'a'b' \);
$

Any string is accepted as the second argument to bless\.  Dumper needs
to handle it accordingly\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low

Site configuration information for perl v5.8.4:

Configured by Debian Project at Thu Mar 23 21:57:50 UTC 2006.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
  Platform:
    osname=linux, osvers=2.6.15.4, archname=i386-linux-thread-multi
    uname='linux ninsei 2.6.15.4 #1 smp preempt mon feb 20 09:48:53 pst 2006 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.4 -Dsitearch=/usr/local/lib/perl/5.8.4 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.4 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3.5 (Debian 1:3.3.5-13)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so.5.8.4
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.4:
    /etc/perl
    /usr/local/lib/perl/5.8.4
    /usr/local/share/perl/5.8.4
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.4:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_GB
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/pub/i686-pc-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/usr/X11R6/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From a.r.ferreira@gmail.com

This tentative patch fix the following Data&#8203;::Dumper behavior&#8203;:

$ cat t0\.pl
use Data&#8203;::Dumper;
print Dumper\(bless\(\{\}\, "a'b"\)\);
$ perl t0\.pl
$VAR1 = bless\( \{\}\, 'a'b' \);

producing instead&#8203;:
$ perl t0\.pl
$VAR1 = bless\( \{\}\, 'a\\'b' \);

Two issues remain&#8203;:

\* should \( Test&#8203;::More => 0\.60 \) be entered in Makefile\.PL ?
&nbsp;  \(the regression test "t/bless\.t" makes use of this requirement\)
&nbsp;  or should I skip?

\* To tell the truth\, I only tried building Data&#8203;::Dumper from outside
the Perl core \(but with 5\.9\.5\)\. But I could not until I overrid the
DEFINE="\-DUSE\_PPORT\_H" in Makefile\.PL

Is it a problem with my installation or something that is outdated?

$ perl Makefile\.PL
Writing Makefile for Data&#8203;::Dumper

$ make
cp Dumper\.pm blib/lib/Data/Dumper\.pm
/home/Ferreira/local/bin/perl5\.9\.5\.exe
/home/Ferreira/local/lib/perl5/5\.9\.5/ExtUtils/xsubpp  \-typemap
/home/Ferreira/local/lib/perl5/5\.9\.5/ExtUtils/typemap  Dumper\.xs >
Dumper\.xsc && mv Dumper\.xsc Dumper\.c
gcc \-c   \-DPERL\_USE\_SAFE\_PUTENV \-U\_\_STRICT\_ANSI\_\_ \-fno\-strict\-aliasing
\-pipe \-I/usr/local/include \-DUSEIMPORTLIB \-O2   \-DVERSION=\\"2\.121\_14\\"
\-DXS\_VERSION=\\"2\.121\_14\\"
"\-I/home/Ferreira/local/lib/perl5/5\.9\.5/cygwin/CORE"  \-DUSE\_PPPORT\_H
Dumper\.c
Dumper\.xs&#8203;:6&#8203;:22&#8203;: ppport\.h&#8203;: No such file or directory \<\<\<\<\< HERE
make&#8203;: \*\*\* \[Dumper\.o\] Error 1

Regards\,
Adriano Ferreira

**Comment by p5pRT** [comment]
### From a.r.ferreira@gmail.com

[dd.diff](https://rt-archive.perl.org/perl5/Ticket/Attachment/275747/120997/dd.diff)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @smpeters

On Tue\, Jul 24\, 2007 at 04&#8203;:29&#8203;:21PM \-0300\, Adriano Ferreira wrote&#8203;:
> This tentative patch fix the following Data&#8203;::Dumper behavior&#8203;:
> 
> $ cat t0\.pl
> use Data&#8203;::Dumper;
> print Dumper\(bless\(\{\}\, "a'b"\)\);
> $ perl t0\.pl
> $VAR1 = bless\( \{\}\, 'a'b' \);
> 
> producing instead&#8203;:
> $ perl t0\.pl
> $VAR1 = bless\( \{\}\, 'a\\'b' \);
> 
> Two issues remain&#8203;:
> 
> \* should \( Test&#8203;::More => 0\.60 \) be entered in Makefile\.PL ?
>  \(the regression test "t/bless\.t" makes use of this requirement\)
>  or should I skip?
> 
> \* To tell the truth\, I only tried building Data&#8203;::Dumper from outside
> the Perl core \(but with 5\.9\.5\)\. But I could not until I overrid the
> DEFINE="\-DUSE\_PPORT\_H" in Makefile\.PL
> 
> Is it a problem with my installation or something that is outdated?
> 
> $ perl Makefile\.PL
> Writing Makefile for Data&#8203;::Dumper
> 
> $ make
> cp Dumper\.pm blib/lib/Data/Dumper\.pm
> /home/Ferreira/local/bin/perl5\.9\.5\.exe
> /home/Ferreira/local/lib/perl5/5\.9\.5/ExtUtils/xsubpp  \-typemap
> /home/Ferreira/local/lib/perl5/5\.9\.5/ExtUtils/typemap  Dumper\.xs >
> Dumper\.xsc && mv Dumper\.xsc Dumper\.c
> gcc \-c   \-DPERL\_USE\_SAFE\_PUTENV \-U\_\_STRICT\_ANSI\_\_ \-fno\-strict\-aliasing
> \-pipe \-I/usr/local/include \-DUSEIMPORTLIB \-O2   \-DVERSION=\\"2\.121\_14\\"
> \-DXS\_VERSION=\\"2\.121\_14\\"
> "\-I/home/Ferreira/local/lib/perl5/5\.9\.5/cygwin/CORE"  \-DUSE\_PPPORT\_H
> Dumper\.c
> Dumper\.xs&#8203;:6&#8203;:22&#8203;: ppport\.h&#8203;: No such file or directory \<\<\<\<\< HERE
> make&#8203;: \*\*\* \[Dumper\.o\] Error 1
> 

In the Perl core\, the ppport\.h files have been removed for the most part\.  That
would explain the compile error\.

Steve Peters
steve@&#8203;fisharerojo\.org

**Comment by p5pRT** [comment]
### From @smpeters

On Tue Jul 24 12&#8203;:30&#8203;:10 2007\, ferreira2 wrote&#8203;:
> This tentative patch fix the following Data&#8203;::Dumper behavior&#8203;:
> 
> $ cat t0\.pl
> use Data&#8203;::Dumper;
> print Dumper\(bless\(\{\}\, "a'b"\)\);
> $ perl t0\.pl
> $VAR1 = bless\( \{\}\, 'a'b' \);
> 
> producing instead&#8203;:
> $ perl t0\.pl
> $VAR1 = bless\( \{\}\, 'a\\'b' \);
> 
> Two issues remain&#8203;:
> 
> \* should \( Test&#8203;::More => 0\.60 \) be entered in Makefile\.PL ?
>   \(the regression test "t/bless\.t" makes use of this requirement\)
>   or should I skip?
> 
> \* To tell the truth\, I only tried building Data&#8203;::Dumper from outside
> the Perl core \(but with 5\.9\.5\)\. But I could not until I overrid the
> DEFINE="\-DUSE\_PPORT\_H" in Makefile\.PL
> 

These really aren't an issue unless someone actually makes a release of
the module\, in which case\, we can sync up at that time\.

Thanks for the patch though\.  I tested it and everything seems fine\.  I
applied it as change \#31651\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From nicholasrperez@gmail.com

Wait\, isn't the single quote behavior valid for backward compatibility
with perl4? Or has that finally been fully obsoleted?

On 7/24/07\, Steve Peters via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Tue Jul 24 12&#8203;:30&#8203;:10 2007\, ferreira2 wrote&#8203;:
> > This tentative patch fix the following Data&#8203;::Dumper behavior&#8203;:
> >
> > $ cat t0\.pl
> > use Data&#8203;::Dumper;
> > print Dumper\(bless\(\{\}\, "a'b"\)\);
> > $ perl t0\.pl
> > $VAR1 = bless\( \{\}\, 'a'b' \);
> >
> > producing instead&#8203;:
> > $ perl t0\.pl
> > $VAR1 = bless\( \{\}\, 'a\\'b' \);
> >
> > Two issues remain&#8203;:
> >
> > \* should \( Test&#8203;::More => 0\.60 \) be entered in Makefile\.PL ?
> >   \(the regression test "t/bless\.t" makes use of this requirement\)
> >   or should I skip?
> >
> > \* To tell the truth\, I only tried building Data&#8203;::Dumper from outside
> > the Perl core \(but with 5\.9\.5\)\. But I could not until I overrid the
> > DEFINE="\-DUSE\_PPORT\_H" in Makefile\.PL
> >
>
> These really aren't an issue unless someone actually makes a release of
> the module\, in which case\, we can sync up at that time\.
>
> Thanks for the patch though\.  I tested it and everything seems fine\.  I
> applied it as change \#31651\.
>

**Comment by p5pRT** [comment]
### From jns@gellyfish.com

On Tue\, 2007\-07\-24 at 17&#8203;:28 \-0500\, Nicholas Perez wrote&#8203;:
> Wait\, isn't the single quote behavior valid for backward compatibility
> with perl4? Or has that finally been fully obsoleted?
> 

No the apostrophe still works as a package separator\, however I think
the patch was aimed at fixing a potential error due to an unmatched
single quote\. 

There is one weirdness with this which I only noticed because I didn't
think this fix was necessary because I thought blessing something as a'b
would always be identical to blessing it into a&#8203;::b \(and consequently you
will always have a&#8203;::b by the time Data&#8203;::Dumper got to it\,\) however this
isn't always the case \- the "'" is only changed to a '&#8203;::' if a'b has
been mentioned as a bareword \(say in a package statement\) somewhere in
the code \- e\.g the following with the "package" line commented and
uncommented&#8203;:

&nbsp;        my $bar = bless \{\}\, 'a\\'b';
&nbsp;        print $bar;
&nbsp;        
&nbsp;        package a'b;

It also works fine if you use a bareword a'b in the bless without the
package but obviously "strict" doesn't like that\.

Bug or feature? Dunno\, but it could potentially take someone by
surprise\.

/J\\
&nbsp;        
&nbsp;        

> On 7/24/07\, Steve Peters via RT \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> > On Tue Jul 24 12&#8203;:30&#8203;:10 2007\, ferreira2 wrote&#8203;:
> > > This tentative patch fix the following Data&#8203;::Dumper behavior&#8203;:
> > >
> > > $ cat t0\.pl
> > > use Data&#8203;::Dumper;
> > > print Dumper\(bless\(\{\}\, "a'b"\)\);
> > > $ perl t0\.pl
> > > $VAR1 = bless\( \{\}\, 'a'b' \);
> > >
> > > producing instead&#8203;:
> > > $ perl t0\.pl
> > > $VAR1 = bless\( \{\}\, 'a\\'b' \);
> > >
> > > Two issues remain&#8203;:
> > >
> > > \* should \( Test&#8203;::More => 0\.60 \) be entered in Makefile\.PL ?
> > >   \(the regression test "t/bless\.t" makes use of this requirement\)
> > >   or should I skip?
> > >
> > > \* To tell the truth\, I only tried building Data&#8203;::Dumper from outside
> > > the Perl core \(but with 5\.9\.5\)\. But I could not until I overrid the
> > > DEFINE="\-DUSE\_PPORT\_H" in Makefile\.PL
> > >
> >
> > These really aren't an issue unless someone actually makes a release of
> > the module\, in which case\, we can sync up at that time\.
> >
> > Thanks for the patch though\.  I tested it and everything seems fine\.  I
> > applied it as change \#31651\.
> >
